client/client.go

```diff
@@ -0,0 +1,462 @@
+// Copyright Contributors to the Open Cluster Management project
+
+package client
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	apimachinerymeta "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+	"sigs.k8s.io/controller-runtime/pkg/ratelimiter"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+var (
+	ErrNotStarted   error = errors.New("DynamicWatcher must be started to perform this action")
+	ErrInvalidInput error = errors.New("invalid input provided")
+)
+
+type Reconciler interface {
+	// Reconcile is called whenever an object has started being watched (if it exists) as well as when it has changed
+	// (added, modified, or deleted). If an error is returned, the request is requeued.
+	Reconcile(ctx context.Context, watcher ObjectIdentifier) (reconcile.Result, error)
+}
+
+// Options are the arguments for creating a new DynamicWatcher.
+type Options struct {
+	// RateLimiter is used to limit how frequently requests may be queued.
+	// Defaults to client-go's MaxOfRateLimiter (like controller-runtime does) which has both overall and per-item rate
```

Comment:
An unnecessary comparison, especially if `controller-runtime` changes but this library doesn't
```suggestion
	// Defaults to client-go's MaxOfRateLimiter which has both overall and per-item rate
```